### PR TITLE
fixed different text box behavior with selections.

### PR DESCRIPTION
**Pull request by columna1**
pressing right when you selection ends at the end of the text clears the selection now
same with pressing left at the beginning of the text

ctrl+selection fixes:
ctrl+move selected text without holding shift doesn't move in weird ways anymore

### CODE REVIEW

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 1]
```diff
@@ -382,33 +383,47 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
...
                     {
                         int lastSpace = text.LastIndexOf(' ', Math.Max(0, selectionEnd - 2));
                         if (lastSpace >= 0)
-                            amount = selectionEnd - lastSpace - 1;
-                        else
+                        {
+                            if (!state.Keyboard.ShiftPressed)
+                                resetSelection();
+                            amount = (selectionEnd - lastSpace - 1);
+                        } else
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: else should be on a new line.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 2]
```diff
@@ -382,33 +383,47 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
...
                             amount = selectionEnd;
                     }
-
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: why remove this?

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 3]
```diff
@@ -403,7 +403,7 @@ protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
                     }
                     moveSelection(-amount, state.Keyboard.ShiftPressed);
                 }
-                return true;
+                    return true;
```
Comment (SirCmpwn) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: :-1: just move the return statements into the { }  if you want to do this.

**Comment by peppy** [comment]
You seem to have added similar code in multiple places. Why don't you just change the behaviour inside `resetSelection` in that case?

also this will probably need commenting. i'm not sure what you're trying to achieve with these shift checks.

File: osu.Framework/Graphics/UserInterface/TextBox.cs (hunk 3 above)
Comment (columna1) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: I'm just trying to keep it the same as it was before

**Comment by SirCmpwn** [comment]
>also this will probably need commenting. i'm not sure what you're trying to achieve with these shift checks.

They make sense to me FWIW. Just read through the code and think about the beavior you're used to seeing on OS text boxes. In the Right handler, for instance, pressing right with shift will increase the selection, but without shift will empty the selection and move the caret alone.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 4]
```diff
@@ -29,6 +29,9 @@ public class TextBox : Container
         public int? LengthLimit;
 
         public bool AllowClipboardExport => true;
+        
+        private bool doubleClickDragging = false;
+        private Vector2 doubleClickWord;
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: remove boolean and make doubleClickWord nullable. also add a comment that this is representing left and right selection coordinates (and is not a positional vector).

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 5]
```diff
@@ -510,13 +540,39 @@ protected override bool OnDrag(InputState state)
...
+                if (getCharacterClosestTo(state.Mouse.Position) > doubleClickWord[1]) 
+                {
+                    selectionEnd = text.IndexOf(' ', getCharacterClosestTo(state.Mouse.Position));
+                    if (selectionEnd < selectionStart)
+                        selectionEnd = text.Length;
+                    selectionStart = doubleClickWord[0];
+                }
+                else if (getCharacterClosestTo(state.Mouse.Position) < doubleClickWord[0]) 
+                {
+                    selectionEnd = text.LastIndexOf(' ', getCharacterClosestTo(state.Mouse.Position))+1;
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: `) + 1`

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 6]
```diff
@@ -576,6 +636,13 @@ protected override bool OnMouseDown(InputState state, MouseDownEventArgs args)
             return true;
         }
 
+        protected override bool OnMouseUp(InputState state, MouseUpEventArgs args)
+        {
+            if (doubleClickWord != null)
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: no need to check for null here

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 7]
```diff
@@ -527,7 +583,7 @@ protected override bool OnDragStart(InputState state)
         }
 
         protected override bool OnDoubleClick(InputState state)
-        {
+        {
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: ?

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 8]
```diff
@@ -82,14 +85,24 @@ public override void Load()
...
         }
 
         private void resetSelection()
         {
-            selectionStart = selectionEnd = text.Length;
+            selectionStart = selectionEnd;
             cursorAndLayout.Invalidate();
         }
 
+        private void resetSelection(bool reset)
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: just remove the bool and do the conditional at the calling site. this is just wrong.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 9]
```diff
@@ -82,14 +85,24 @@ public override void Load()
 
             textContainer.Add(cursor);
             textContainer.Add(textFlow);
+
```
Comment (peppy) [inline osu.Framework/Graphics/UserInterface/TextBox.cs]: remove empty line

**Comment by peppy** [comment]
nice job 👍

**Pull request merged**